docs/subnets/create-a-vm-blobvm.md

```diff
@@ -0,0 +1,1353 @@
+# How to Build a Complex VM
+
+This is part of a series of tutorials for building a Virtual Machine (VM):
+
+- [Introduction to Virtual Machines](./introduction-to-vm.md)
+- [How to Build a Simple VM](./create-a-vm-timestampvm.md)
+- How to Build a Complex VM (this article)
+
+## Introduction
+
+In this tutorial, we will learn all about Virtual Machines and how we can build them by taking references from [BlobVM](https://github.com/ava-labs/blobvm). It's a Virtual Machine that can be used to instantiate key-value blockchains for storing files like images, videos, etc. in an efficient way.
+
+Blobs are small chunks of data. In BlobVM, we divide a file into small blobs and store them as a key-value pair. The key of these small chunks of a file is linked together as the children of the root. This tree is also stored as JSON data against the root's key.
+
+## Prerequisites
+
+Make sure you have followed the previous tutorials in this series:
+
+- [Introduction to Virtual Machines](./introduction-to-vm.md)
+- [How to Build a Simple VM](./create-a-vm-timestampvm.md)
+
+## Components of the BlobVM
+
+The main task of a VM is to represent initial state (genesis state) and block structure containing state transitioning details. One of the most common structure to represent state transitions is **transaction**. When a block with transactions, is applied to the current state, the state transition will happen by simply executing the transactions. The order of transactions will matter here.
+
+BlobVM has the following components to handle the tasks from transaction to block acceptance:
+
+- **Transaction** - Transaction structure, initialization, execution etc.
+- **Mempool** - Heap (min and max) for keeping pending transactions locally on the node.
+- **Network** - Implements gossip of transactions from mempool to the network.
+- **Block** - Handles block creation, initialization, verification, parsing, etc.
+- **Builder** - Returns the built preferred block by including transactions from the mempool.
+- **Block Builder** - Initiates the gossip of new transactions, notifies the engine that new blocks are ready to be built, and also handles next block generation notification.
+- **Storage** - Stores and retrieves data of the chain's state.
+- **Virtual Machine** - Entry point for all the components to orchestrate them according to the consensus engine and API requests.
+- **Service** - API handlers for interacting with VM and initialized chain.
+- **Factory** - For creating new instances of the Virtual Machine.
+
+## Transaction Lifecycle in BlobVM
+
+A Virtual Machine exposes APIs or handlers for users to make direct RPC to service or use the client to interact with the service. Every change on a chain happens through blocks and more specifically, transactions. A VM handles transactions internally since the consensus engine only cares about the block. Let's see, how a transaction goes through the network to update the chain's state:
```

Comment:
what does the client in `use the client` refer to?

Reply:
I should remove the `client` reference, as this is not a requirement for BlobVM.